Source/AwsCommonRuntimeKit/crt/Future.swift

```diff
@@ -0,0 +1,76 @@
+//  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//  SPDX-License-Identifier: Apache-2.0.
+
+import Dispatch
+
+public class Future<Value> {
+    public typealias FutureResult = Result<Value, Error>
+    
+    private var _value: FutureResult? {
+        get {
+            queue.sync { [weak self] in
+                guard let strongSelf = self else { return nil }
+                return strongSelf._value
+            }
+        }
+        set(value) {
+            queue.async(flags: .barrier) { [weak self] in
+                guard let strongSelf = self else { return }
+                strongSelf._value = value
+            }
+        }
+    }
+    
+    private let queue = DispatchQueue(label: "atomicValue", qos: .default, attributes: .concurrent)
+    
+    var waiter = DispatchSemaphore(value: 0)
+    
+    private var _observers = [((FutureResult) -> Void)]()
+    
+    public init(value: FutureResult? = nil) {
+        self._value = value
+    }
+    
+    public func get() throws -> Value {
+        if let value = self._value {
+            return try value.get()
+        }
+        
+        waiter.wait()
+        //can force unwrap here cuz if wait was lifted, future has completed
+        return try self._value!.get()
+    }
+    
+    public func complete(_ value: FutureResult) {
+        self._value = value
+        self._observers.forEach { $0(value) }
+        waiter.signal()
+    }
+    
+    public func then(_ block: @escaping (FutureResult) -> Void) {
+        if let value = self._value {
```

Comment:
What if the Value type is Void? Value will most likely be nil/falsy, won't it?

Seems like something we should add unit tests for :)

Reply:
ah good point, let me write one :)